Conch/lib/Conch/Validation.pm

```diff
@@ -0,0 +1,625 @@
+
+	$self->register_result(
+		expected => 'hello',
+		got      => 'hello',
+		message  => 'Hello world!'
+	);
+
+=item C<category>
+
+By default, the validation result stores the C<category> attribute of the
+Validation class. You may override the validation result category with this
+attribute.
+
+	$self->register_result(
+		expected => 'hello',
+		got      => 'hello',
+		category => 'BIOS'
+	);
+
+=item C<component_id>
+
+You may specify the optional string attribute C<component_id> to set an
+identifier to help identify a specific component under test.
+
+	$self->register_result(
+		expected  => 'OK',
+		got       => $disk->{health},
+		hint      => $disk->{serial_number}
+	);
+
+=item C<hint>
+
+You may specify the optional string attribute C<hint> to be stored B<if>
+the validation fails. This string should help identify to a user how to fix the
+validation failure. If the validation succeeds, the hint string is not stored
+in the validation result.
+
+	$self->register_result(
+		expected  => 'hello',
+		got       => 'bye',
+		hint      => "Try saying 'hello' instead"
+	);
+
+=back
+
+=cut
+
+sub register_result ( $self, %attrs ) {
+	my $expected = $attrs{expected};
+	my $got      = $attrs{got};
+	my $cmp_op   = $attrs{cmp} || 'eq';
+
+	$self->die( "'expected' value must be defined", 2 )
+		unless defined($expected);
+
+	return $self->fail("'got' value is undefined") unless defined($got);
+
+	$self->die( "'got' value must be a scalar", 2 ) if ref($got);
+
+	if ( $cmp_op eq 'oneOf' ) {
+		$self->die( "'expected' value must be an array when comparing with 'oneOf'",
+			2 )
+			unless ref($expected) eq 'ARRAY';
+	}
+	elsif ( $cmp_op eq 'like' ) {
+		$self->die(
+			"'expected' value must be a scalar or Regexp when comparing with 'like'",
+			2
+		) unless ref($expected) eq 'Regexp' || ref($expected) eq '';
+	}
+	else {
+		$self->die(
+			"'expected' value must be a scalar when comparing with '$cmp_op'", 2 )
+			if ref($expected);
+	}
+
+	my $success = $cmp_dispatch->{$cmp_op}->( $got, $expected );
+	my $expected_got_message;
+	if ( $cmp_op eq 'oneOf' ) {
+		$expected_got_message =
+			  'Expected one of: '
+			. join( ', ', map { "'$_'" } $expected->@* )
+			. ". Got '$got'.";
+	}
+	else {
+		$expected_got_message = "Expected $cmp_op '$expected'. Got '$got'.";
+	}
+
+	my $validation_result = $self->{_validation_result_builder}->(
+		message  => $attrs{message}  || $expected_got_message,
+		name     => $attrs{name}     || $self->name,
+		category => $attrs{category} || $self->category,
+		status => $success ? STATUS_SUCCESS : STATUS_FAIL,
+		hint   => $success ? $attrs{hint}   : undef
+	);
+
+	push $self->validation_results->@*, $validation_result;
+	return $self;
+
+}
+
+=head2 die
+
+Stop execution of the Validation immediately and record an error.
+
+	$self->die('This validation cannot continue!') if $bad_condition;
+
+=cut
+
+sub die ( $self, $message, $level = 1 ) {
+
+	die Mojo::Exception->new($message)->trace($level);
```

Comment:
My intention is that validations should die fast and loudly. Exceptions are caught and made into 'error' validation results and are reported in the API. I want to minimize ambiguity as much as possible. I want to avoid unexpected undefs cast into empty string in comparisons and validations that run but don't produce validation results because of a conditional branch. Dying fast also helps to be explicit about the inputs when writing the validation unit tests.

I see your point about the methods that immediately die if they have undefined data, like `device_location`, `hardware_product_profile`, etc. I'm not sure will be a problem under normal operation. We can expect hardware product always to be present when validation plans are executed by the orchestration system because it's a necessary component of a workflow. If a workflow step is executed before a device may have a device location, then it should be associated with a validation plan of validations that don't require a location.

I'll go ahead and add a `has_device_location` method to return a boolean in the case that there's some conditional based on whether a device has an assigned location. I don't there will be many validations that use this type of logic (and they might be better split into two validations in separate pre- and post-location plans), but it's possibly useful.

Reply:
See also my comment in https://github.com/joyent/conch/pull/162#discussion_r176791133